toolchain/check/declaration_name_stack.cpp

```diff
@@ -39,7 +40,12 @@ auto DeclarationNameStack::Pop() -> NameContext {
     ApplyNameQualifier(parse_node, name_id);
   }
 
-  return declaration_name_stack_.pop_back_val();
+  return declaration_name_stack_.back();
+}
+
+auto DeclarationNameStack::Pop() -> void {
+  context_->PopToScope(declaration_name_stack_.back().enclosing_scope);
+  declaration_name_stack_.pop_back();
```

Comment:
I think the `DeclarationNameStack` mutating the context when building up the declaration name is the simplest approach; otherwise I think we end up moving a lot of the logic currently in `DeclarationNameStack::UpdateScopeIfNeeded` and its calling logic out to somewhere else. And if `DeclarationNameStack` is pushing the scopes, I think logically it should be popping the scopes too.

Extended the comment to say more about popping the scopes, and updated the comment for `ApplyNameQualifier` to note that it can enter scopes.